### ISSUE TITLE
SerializationException while using Mapper with Unit

### ISSUE THREAD

**Issue by juggernaut0** [Kotlin/kotlinx.serialization#647]
**Describe the bug**
Attempting to use Mapper to map Unit (using UnitSerializer) fails with the following exception:

```
kotlinx.serialization.SerializationException: No tag in stack for requested element
	at kotlinx.serialization.TaggedEncoder.popTag(Tagged.kt:132)
	at kotlinx.serialization.TaggedEncoder.encodeUnit(Tagged.kt:66)
	at kotlinx.serialization.internal.UnitSerializer.serialize(Primitives.kt:37)
	at kotlinx.serialization.internal.UnitSerializer.serialize(Primitives.kt:34)
	at kotlinx.serialization.Encoder$DefaultImpls.encodeSerializableValue(Coders.kt:29)
	at kotlinx.serialization.TaggedEncoder.encodeSerializableValue(Tagged.kt:15)
	at kotlinx.serialization.CoreKt.encode(Core.kt:74)
	at kotlinx.serialization.Mapper.map(Mapper.kt:101)
	at kotlinx.serialization.Mapper$Companion.map(Mapper.kt:138)
	at UnitMapperTest.mapUnit(UnitMapperTest.kt:9)
```

mapNullable throws a similar exception.

**To Reproduce**
```kotlin
class UnitMapperTest {
    @Test
    fun mapUnit() {
        assertEquals(emptyMap(), Mapper.map(UnitSerializer, Unit))
    }
}
```

**Expected behavior**
The above test should pass. I.E. Unit should be able to be mapped to an emptyMap.

**Environment**
- Kotlin version: 1.3.61
- Library version: 0.14.0
- Kotlin platforms: JVM, JS
- Gradle version: 6.0.1

**Comment by sandwwraith** [comment]
It's not clear why `Unit` is an empty map, actually. We treat Unit as a primitive type, like Int, and `Mapper` does not work with an Int either.

**Comment by juggernaut0** [comment]
Like I mentioned on https://github.com/Kotlin/kotlinx.serialization/issues/645 it is useful in generic situations to have a type that represents "No Data" or "Empty Map", and I was hoping to leverage Unit for this because at first glance it is a class with no fields, its JSON representation appears to be an empty map, and its meaning is already well-known as "No Return Data".

However you seem to be suggesting that this is coincidence and that Unit should be treated as primitive-like.

`kotlin.collections.EmptyMap` would be a useful alternative but unfortunately it is private.

**Comment by sandwwraith** [comment]
On current `dev` this test passes. I think it got fixed automatically with #664
